<h2>Classification Using Ecological Dataset</h2>


<h5>Importing the dataset</h5>

In [3]:
import pandas as pd
dataset=pd.read_csv('ecology_dataset.csv')
dataset

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
5     6.999209  9.109247  3.784066  4.267788  4.429669 -1.857139 -0.670397   
6     6.710166  3.765204  6.929314  8.818562  2.397419 -0.614590 -1.208826   
7     6.953512  1.379125  5.719400  7.870307  3.224495 -0.748998 -1.186517   
8     4.689852  4.007747  1.478573  3.733787  4.041300 -1.410344 -1.238204   
9     9.841496  1.413822  9.769856  7.641616  4.727595 -1.991363 -0.857637   
10    5.930110  6.730873  6.245138  0.533288  2.327092 -0.702501 -1.116920   
11    5.381299  8.014521  8.095174  6.769248  5.507551 -1.972714 -1.849333   
12    1.616787  2.939228  0.819791  4.191804  3.752282 -1.484885 -1.280581   
13    8.551598  8.314952  2.549964  9.926807  4.891714 -1.808626 -1.167061   
14    1.132108  2.920324  8.951079  7.248583  5.033681 -1.846080 -1.362777   
15    7.021362  4.374294  4.775904  8.838426  3.335857 -0.962392 -1.407627   
16    4.952241  8.088672  8.883319  5.694557  5.067296 -1.681414 -1.877061   
17    4.142830  2.439089  1.290456  9.456443  3.934796 -1.469299 -1.766941   
18    9.346126  7.920030  2.335276  3.269181  4.581174 -1.106753 -1.747078   
19    3.931954  9.180890  6.064480  6.292147  5.363996 -1.695085 -1.880273   
20    8.676895  5.583325  4.925402  7.356212  2.828247 -1.396732 -0.929290   
21    7.383177  5.253173  0.924517  0.744899  3.976105 -1.429219 -1.424951   
22    4.541829  1.969862  8.513193  7.562036  4.362835 -1.626414 -1.761731   
23    5.973186  6.043118  5.996045  1.076940  3.828218 -0.989549 -1.079026   
24    7.253346  7.733517  1.621091  7.890610  3.567240 -1.377855 -1.449532   
25    6.192481  2.533877  1.662380  5.163924  3.462392 -1.355954 -1.432846   
26    5.895693  1.096302  1.311732  8.777868  4.069170 -1.063139 -1.374427   
27    0.727077  6.425554  6.269094  0.786911  2.913179 -1.187866 -0.687735   
28    1.807819  7.020356  6.676929  5.169648  3.672288 -1.188471 -0.819164   
29    9.034147  1.071635  5.478342  9.412525  3.405139 -1.163253 -1.653954   
...        ...       ...       ...       ...       ...       ...       ...   
9970  5.137851  0.737253  5.725703  9.068354  3.499583 -1.186086 -1.326108   
9971  0.619440  4.882592  3.360420  4.208011  3.981196 -1.438492 -1.837809   
9972  9.694989  8.393842  4.536673  5.012365  3.988816 -0.837072 -1.525764   
9973  5.485364  3.096149  8.925530  2.090592  4.365430 -1.888132 -1.574179   
9974  0.960580  3.984473  6.134353  9.755264  4.076419 -1.425252 -1.523594   
9975  7.078400  4.660627  3.535984  2.153805  3.055565 -0.988581 -0.899321   
9976  7.063097  3.897790  3.667961  7.953410  4.632592 -1.146609 -1.850362   
9977  3.749501  3.613827  1.868883  2.435817  2.410419 -0.605413 -0.837323   
9978  8.632397  4.239094  1.240248  9.246927  4.092402 -0.868173 -1.439817   
9979  7.512814  9.616251  8.045889  2.200809  2.377325 -0.688485 -0.703106   
9980  4.141337  1.112868  8.400888  1.869624  3.671563 -0.699052 -1.513896   
9981  5.984990  7.833828  4.670350  8.941526  3.562268 -1.505071 -0.891233   
9982  6.756528  7.107648  6.038633  8.236088  1.980384 -0.799625 -0.642624   
9983  1.482214  3.121273  5.220920  1.609288  4.394557 -1.924542 -0.881996   
9984  2.501787  3.087194  7.493896  2.177944  3.817471 -0.958986 -1.044281   
9985  1.976306  6.894304  9.347806  5.763746  3.403860 -1.880294 -0.589551   
9986  6.038326  5.621908  5.099137  2.279340  3.013139 -1.347185 -0.585465   
9987  3.176332  5.101491  8.134258  4.985496  3.031084 -0.957564 -1.102893   
9988  3.871136  7.371163  2.579495  1.540674  3.492453 -0.635868 -1.256279   
9989  9.929407  0.863541  8.1611

<p>Checking out for missing values and duplicated rows and dropping the unwanted column</p>

In [4]:
dataset.drop(['stab'], axis=1)

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
5     6.999209  9.109247  3.784066  4.267788  4.429669 -1.857139 -0.670397   
6     6.710166  3.765204  6.929314  8.818562  2.397419 -0.614590 -1.208826   
7     6.953512  1.379125  5.719400  7.870307  3.224495 -0.748998 -1.186517   
8     4.689852  4.007747  1.478573  3.733787  4.041300 -1.410344 -1.238204   
9     9.841496  1.413822  9.769856  7.641616  4.727595 -1.991363 -0.857637   
10    5.930110  6.730873  6.245138  0.533288  2.327092 -0.702501 -1.116920   
11    5.381299  8.014521  8.095174  6.769248  5.507551 -1.972714 -1.849333   
12    1.616787  2.939228  0.819791  4.191804  3.752282 -1.484885 -1.280581   
13    8.551598  8.314952  2.549964  9.926807  4.891714 -1.808626 -1.167061   
14    1.132108  2.920324  8.951079  7.248583  5.033681 -1.846080 -1.362777   
15    7.021362  4.374294  4.775904  8.838426  3.335857 -0.962392 -1.407627   
16    4.952241  8.088672  8.883319  5.694557  5.067296 -1.681414 -1.877061   
17    4.142830  2.439089  1.290456  9.456443  3.934796 -1.469299 -1.766941   
18    9.346126  7.920030  2.335276  3.269181  4.581174 -1.106753 -1.747078   
19    3.931954  9.180890  6.064480  6.292147  5.363996 -1.695085 -1.880273   
20    8.676895  5.583325  4.925402  7.356212  2.828247 -1.396732 -0.929290   
21    7.383177  5.253173  0.924517  0.744899  3.976105 -1.429219 -1.424951   
22    4.541829  1.969862  8.513193  7.562036  4.362835 -1.626414 -1.761731   
23    5.973186  6.043118  5.996045  1.076940  3.828218 -0.989549 -1.079026   
24    7.253346  7.733517  1.621091  7.890610  3.567240 -1.377855 -1.449532   
25    6.192481  2.533877  1.662380  5.163924  3.462392 -1.355954 -1.432846   
26    5.895693  1.096302  1.311732  8.777868  4.069170 -1.063139 -1.374427   
27    0.727077  6.425554  6.269094  0.786911  2.913179 -1.187866 -0.687735   
28    1.807819  7.020356  6.676929  5.169648  3.672288 -1.188471 -0.819164   
29    9.034147  1.071635  5.478342  9.412525  3.405139 -1.163253 -1.653954   
...        ...       ...       ...       ...       ...       ...       ...   
9970  5.137851  0.737253  5.725703  9.068354  3.499583 -1.186086 -1.326108   
9971  0.619440  4.882592  3.360420  4.208011  3.981196 -1.438492 -1.837809   
9972  9.694989  8.393842  4.536673  5.012365  3.988816 -0.837072 -1.525764   
9973  5.485364  3.096149  8.925530  2.090592  4.365430 -1.888132 -1.574179   
9974  0.960580  3.984473  6.134353  9.755264  4.076419 -1.425252 -1.523594   
9975  7.078400  4.660627  3.535984  2.153805  3.055565 -0.988581 -0.899321   
9976  7.063097  3.897790  3.667961  7.953410  4.632592 -1.146609 -1.850362   
9977  3.749501  3.613827  1.868883  2.435817  2.410419 -0.605413 -0.837323   
9978  8.632397  4.239094  1.240248  9.246927  4.092402 -0.868173 -1.439817   
9979  7.512814  9.616251  8.045889  2.200809  2.377325 -0.688485 -0.703106   
9980  4.141337  1.112868  8.400888  1.869624  3.671563 -0.699052 -1.513896   
9981  5.984990  7.833828  4.670350  8.941526  3.562268 -1.505071 -0.891233   
9982  6.756528  7.107648  6.038633  8.236088  1.980384 -0.799625 -0.642624   
9983  1.482214  3.121273  5.220920  1.609288  4.394557 -1.924542 -0.881996   
9984  2.501787  3.087194  7.493896  2.177944  3.817471 -0.958986 -1.044281   
9985  1.976306  6.894304  9.347806  5.763746  3.403860 -1.880294 -0.589551   
9986  6.038326  5.621908  5.099137  2.279340  3.013139 -1.347185 -0.585465   
9987  3.176332  5.101491  8.134258  4.985496  3.031084 -0.957564 -1.102893   
9988  3.871136  7.371163  2.579495  1.540674  3.492453 -0.635868 -1.256279   
9989  9.929407  0.863541  8.1611

In [4]:
dataset.isnull().sum()

tau1     0
tau2     0
tau3     0
tau4     0
p1       0
p2       0
p3       0
p4       0
g1       0
g2       0
g3       0
g4       0
stab     0
stabf    0
dtype: int64

<p>From the results shown above, it is apparent that the dataset is devoid of missing values and duplicated values</p>

<p>Data Preprocessing</p>

In [5]:
x=dataset.iloc[:,0:12].values
y=dataset.iloc[:,13:14].values

In [24]:
#splitting the dataset into the ration 80:20
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1)


In [25]:
#transforming the dataset using standard scaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit_transform(x_train,y_train)
scaler.transform(x_test)

array([[ 0.59395058, -0.41273345,  1.50392381, ...,  1.1672034 ,
        -1.50732963,  1.08472557],
       [ 0.2021896 ,  0.37441634, -0.18880047, ..., -0.39566024,
         1.41465051,  1.22601069],
       [-1.079044  , -0.31374544, -0.88463426, ..., -1.43849538,
         0.65182081, -1.6821675 ],
       ...,
       [ 0.94782488, -1.66372653, -1.65391963, ...,  0.12639128,
         0.57344494,  1.31934985],
       [-1.1202346 ,  0.19397855, -0.2378051 , ...,  0.79408717,
        -1.36232268, -0.80197116],
       [-1.37764025,  1.51186671,  0.28265058, ..., -0.91749729,
         0.00295027,  1.18902334]])

<h4>Random Forest Classifier</h4>

In [18]:
from numpy import mean
from numpy import std
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.ensemble import RandomForestClassifier
# define the model
model = RandomForestClassifier()
model.fit(x_train,y_train)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [30]:
# evaluating the model
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
n_scores = cross_val_score(model, x_test, y_test, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))

Accuracy: 0.847 (0.026)


<p>This explains that a classification accuracy of about 84.7% has been achieved using the random forest classifier</p>

<h4>Extra Trees Classifier</h4>

In [38]:
# evaluate extra trees algorithm for classification
from numpy import mean
from numpy import std
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.ensemble import ExtraTreesClassifier
# define the model
xmodel = ExtraTreesClassifier()
xmodel.fit(x_train,y_train)

C:\Users\Faheeda\Documents\Python\New folder\lib\site-packages\ipykernel_launcher.py:9: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  if __name__ == '__main__':


ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [39]:
# evaluating the model using repeated stratified K_fold
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
n_scores = cross_val_score(xmodel, x_test, y_test, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy: %.4f (%.4f)' % (mean(n_scores), std(n_scores)))

Accuracy: 0.8405 (0.0258)


<p>Just like Random Forest, the accuracy observed when using extra classifiers is 84.1% </p>

<h4>XGBoost</h4>

In [42]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score

In [43]:
#fitting the XGboost to the training set
xgmodel = XGBClassifier()
xgmodel.fit(x_train, y_train)

C:\Users\Faheeda\Documents\Python\New folder\lib\site-packages\sklearn\preprocessing\label.py:95: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Faheeda\Documents\Python\New folder\lib\site-packages\sklearn\preprocessing\label.py:128: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0, learning_rate=0.1,
       max_delta_step=0, max_depth=3, min_child_weight=1, missing=None,
       n_estimators=100, n_jobs=1, nthread=None,
       objective='binary:logistic', random_state=0, reg_alpha=0,
       reg_lambda=1, scale_pos_weight=1, seed=None, silent=None,
       subsample=1, verbosity=1)

In [51]:
#Evaluating the model
y_pred = xgmodel.predict(x_test)
# evaluate predictions
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

Accuracy: 91.95%


C:\Users\Faheeda\Documents\Python\New folder\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


<p>The observed accuracy from the XGBoost model is 91.95%, and this is better.</p>

<h4>Randomized Search Cross validation</h4>

In [55]:
#Improving the extra classifiers using hyperparameter tuning
from sklearn.model_selection import RandomizedSearchCV
rsmodel=ExtraTreesClassifier()
n_estimators= [50, 100, 300, 500, 1000]
min_samples_split = [2, 3, 5, 7, 9]
min_samples_leaf = [1, 2, 4, 6, 8]
max_features = ['auto', 'sqrt', 'log2', None] 
hyperparameter_grid = {'n_estimators': n_estimators,
                       'min_samples_leaf': min_samples_leaf,
                       'min_samples_split': min_samples_split,
                       'max_features': max_features}
rscv=RandomizedSearchCV(rsmodel,hyperparameter_grid)

In [ ]:
best_model=rscv.fit(x_train,y_train)
best_model.best_estimator_

<p>The model above will produce the best precision since it has been iterated severally on different subsets of data </p>